### ISSUE TITLE
Clippy suggests replacing `.or_insert(Default::default())` with `.or_default()` inside implementation of `.or_default()`

### ISSUE THREAD

**Issue by schuelermine** [rust-lang/rust-clippy#10228]
### Summary

Clippy suggests replacing `.or_insert(Default::default())` with `.or_default()` inside implementation of `.or_default()`

### Lint Name

or_fun_call

### Reproducer

I tried this code:

```rust
enum MyEntry<'a, T> {
    Occupied(&'a mut T),
    Vacant(&'a mut T),
}

impl<'a, T> MyEntry<'a, T> {
    fn or_insert(self, default: T) {
        match self {
            MyEntry::Occupied(_) => {}
            MyEntry::Vacant(v) => *v = default,
        }
    }
}

impl<'a, T: Default> MyEntry<'a, T> {
    fn or_default(self) {
        self.or_insert(Default::default())
    }
}
```

I saw this happen:

```
warning: use of `or_insert` followed by a call to `default`
   --> src/***.rs:***:14
    |
*** |         self.or_insert(Default::default())
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `or_default()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
```

I expected to see this happen:

Nothing

### Version

```text
rustc 1.65.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 14.0.6
```
